### PR TITLE
Add the ability validate changes to device update configuration make sense

### PR DESCRIPTION
**Pull request by doanac**

### CODE REVIEW

**Comment by doanac** [comment]
This is a bit of a work in progress we should experiment with. Here's some examples of what it can do:
## Preventing a downgrade:
~~~
$ fioctl devices config updates andy-test --dryrun --tags promoted
ERROR: Given tags [promoted] appear to result in a downgrade from version 111 to 80
~~~

## Preventing a bad tag:
~~~
$ fioctl devices config updates andy-test --dryrun --tags promotedd
ERROR: Given tags [promotedd] are not present in targets.json
~~~

## Preventing a bad apps:
~~~
$ fioctl devices config updates andy-test --dryrun --apps foo
ERROR: Given apps [foo] not present in Target version 113 device would update to: [fiotest homeassistant homelab shellhttpd helloworld]
~~~

I'm not sure about the wording of "downgrade". The device wouldn't actually downgrade it would just sort of be purgatory until a new build was performed. We also need to add a "-f" to ignore this and force the change through.

File: subcommands/common.go [hunk 1]
```diff
@@ -164,3 +164,14 @@ func PrintConfigs(deviceConfigs []client.DeviceConfig, listLimit int) int {
 	}
 	return listLimit
 }
+
+func IntersectionInSlices(list1, list2 []string) bool {
```
Comment (mike-sul) [inline subcommands/common.go:168]: That's really interesting to fidn out that golang (base modules) don't include such basic functionality.

File: subcommands/devices/config_updates.go [hunk 2]
```diff
@@ -80,8 +84,38 @@ func loadSotaConfig(device string) *toml.Tree {
...
+		if !match {
+			return false
+		}
+	}
+	return true
+}
+
 // Validate the inputs: Must be alphanumeric, a dash, underscore, or comma
-func validateUpdateArgs() {
+func validateUpdateArgs(device *client.Device) {
```
Comment (mike-sul) [inline subcommands/devices/config_updates.go:118]: Maybe it makes sense to do this validation in the OTA lite API implementation so direct calls to it are validated too?

File: subcommands/devices/config_updates.go [hunk 3]
```diff
@@ -94,6 +128,82 @@ func validateUpdateArgs() {
 		fmt.Println("       apps must be ", pattern)
 		os.Exit(1)
 	}
+
+	tags := device.Tags
+	if len(updateTags) > 0 {
+		tags = strings.Split(updateTags, ",")
+	}
+	apps := device.DockerApps
```
Comment (mike-sul) [inline subcommands/devices/config_updates.go:136]: Will it work for Compose Apps?

Comment (doanac) [reply subcommands/devices/config_updates.go:136]: yes. "docker apps" as a device attribute means what array of apps its reporting up. so its package manager agnostic.

File: subcommands/devices/config_updates.go (hunk 2 above)
Comment (doanac) [reply subcommands/devices/config_updates.go:118]: That's a really interesting idea. I'll look into this today.

**Comment by mike-sul** [comment]
It looks like it's better to have the validation in the OTA Lite API implemenation as we expect that the API will be called not just by fioctl?

**Comment by doanac** [comment]
> It looks like it's better to have the validation in the OTA Lite API implemenation as we expect that the API will be called not just by fioctl?

yes. i think that's way better. This will make it easier to build a GUI that's consistent.

**Comment by doanac** [comment]
I've moved the real checking to happen server side:

 https://git.foundries.io/development/cloudplatforms/infrastructure/ota-lite/merge_requests/64

Once merged, we can clean up and simplify this logic.

**Comment by doanac** [comment]
I've implemented this on the server side.

**Pull request closed without merging**